File: app/src/main/java/com/kksionek/photosyncer/view/ViewPagerAdapter.java
package com.kksionek.photosyncer.view;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import androidx.annotation.NonNull;
import androidx.viewpager.widget.PagerAdapter;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kksionek.photosyncer.R;
import com.kksionek.photosyncer.data.Contact;
import com.kksionek.photosyncer.model.ContactsAdapter;

import io.realm.Realm;
import io.realm.RealmResults;
import io.realm.Sort;

class ViewPagerAdapter extends PagerAdapter {

    private static final int TAB_COUNT = 3;

    private Activity mParentActivity;
    private Realm mRealmUi;

    ViewPagerAdapter(Activity parentActivity, Realm realmUi) {
        mParentActivity = parentActivity;
        mRealmUi = realmUi;
    }

    @Override
    public int getCount() {
        return TAB_COUNT;
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view == object;
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        Context ctx = container.getContext();
        View view = LayoutInflater.from(ctx)
                .inflate(R.layout.tab_not_synced, container, false);
        container.addView(view);

        RecyclerView recyclerView = view.findViewById(R.id.recycler_view);
        RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(ctx);
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.addItemDecoration(
                new DividerItemDecoration(
                        mParentActivity,
                        DividerItemDecoration.VERTICAL));

        ContactsAdapter<Contact> contactsAdapter = null;

        switch (position) {
            case 0: {
                RealmResults<Contact> notSyncedContacts = mRealmUi.where(Contact.class)
                        .isNull("mRelated")
                        .equalTo("mManual", false)
                        .findAll()
                        .sort("mName", Sort.ASCENDING);

                contactsAdapter = new ContactsAdapter<>(mParentActivity, notSyncedContacts, true);
                contactsAdapter.setOnItemClickListener((view1, contact) -> {
                    Intent facebookPicketIntent = new Intent(mParentActivity, FacebookPickerActivity.class);
                    facebookPicketIntent.putExtra(FacebookPickerActivity.EXTRA_ID, contact.getId());
                    mParentActivity.startActivityForResult(facebookPicketIntent, TabActivity.REQUEST_FACEBOOK_PICKER);
                });
                break;
            }
            case 1: {
                RealmResults<Contact> autoSyncedContacts = mRealmUi.where(Contact.class)
                        .isNotNull("mRelated")
                        .equalTo("mManual", false)
                        .findAll()
                        .sort("mName", Sort.ASCENDING);

                contactsAdapter = new ContactsAdapter<>(mParentActivity, autoSyncedContacts, true);
                contactsAdapter.setOnItemClickListener((v, contact) -> {
                    AlertDialog.Builder builder = new AlertDialog.Builder(mParentActivity);
                    builder.setTitle(R.string.alert_cancel_auto_sync_title);
                    builder.setMessage(R.string.alert_cancel_auto_sync_message);
                    builder.setPositiveButton(android.R.string.ok, (dialogInterface, i) -> {
                        mRealmUi.executeTransaction(realm -> {
                            contact.setRelated(null);
                            contact.setManual(true);
                            contact.setSynced(true);
                        });
                        dialogInterface.dismiss();
                    });
                    builder.setNegativeButton(android.R.string.cancel, ((dialogInterface, i) -> dialogInterface.dismiss()));
                    builder.create().show();
                });
                break;
            }
            case 2: {
                RealmResults<Contact> manualContacts = mRealmUi.where(Contact.class)
                        .equalTo("mManual", true)
                        .findAll()
                        .sort("mName", Sort.ASCENDING);

                contactsAdapter = new ContactsAdapter<>(mParentActivity, manualContacts, true);
                contactsAdapter.setOnItemClickListener((v, contact) -> {
                    AlertDialog.Builder builder = new AlertDialog.Builder(mParentActivity);
                    builder.setTitle(R.string.alert_release_bond_title);
                    builder.setMessage(R.string.alert_release_bond_message);
                    //TODO: make another dialog/preference remembering if app should remove photo automatically
                    builder.setPositiveButton(android.R.string.ok, (dialogInterface, i) -> {
                        mRealmUi.executeTransaction(realm -> {
                            contact.setRelated(null);
                            contact.setManual(false);
                            contact.setSynced(false);
                        });
                        dialogInterface.dismiss();
                    });
                    builder.setNegativeButton(android.R.string.cancel, ((dialogInterface, i) -> dialogInterface.dismiss()));
                    builder.create().show();
                });
                break;
            }
        }
        if (contactsAdapter != null)
            recyclerView.setAdapter(contactsAdapter);
        return view;
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        container.removeView((View) object);
    }

    @Override
    public CharSequence getPageTitle(int position) {
        switch (position) {
            case 0:
                return mParentActivity.getString(R.string.activity_tab_tab_not_synced);
            case 1:
                return mParentActivity.getString(R.string.activity_tab_tab_auto);
            case 2:
                return mParentActivity.getString(R.string.activity_tab_tab_manual);
            default:
                return super.getPageTitle(position);
        }
    }
}

File: app/src/main/java/com/kksionek/photosyncer/view/ViewPagerAdapter.kt
package com.kksionek.photosyncer.view

import android.app.Activity
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.viewpager.widget.PagerAdapter
import com.kksionek.photosyncer.R
import com.kksionek.photosyncer.data.Contact
import com.kksionek.photosyncer.model.ContactsAdapter
import io.realm.Realm
import io.realm.Sort
import kotlinx.android.synthetic.main.tab_not_synced.view.*

internal class ViewPagerAdapter(
    private val parentActivity: Activity,
    private val realmUi: Realm
) : PagerAdapter() {

    override fun getCount(): Int = TAB_COUNT

    override fun isViewFromObject(view: View, `object`: Any): Boolean = view === `object`

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val context = container.context
        val view = LayoutInflater.from(context)
            .inflate(R.layout.tab_not_synced, container, false)
        container.addView(view)

        val recyclerView = view.recycler_view
        val layoutManager = LinearLayoutManager(context)
        recyclerView.layoutManager = layoutManager
        recyclerView.addItemDecoration(
            DividerItemDecoration(context, DividerItemDecoration.VERTICAL)
        )

        val contactsAdapter: ContactsAdapter<Contact>

        when (position) {
            0 -> {
                val notSyncedContacts = realmUi.where(Contact::class.java)
                    .isNull("related")
                    .equalTo("isManual", false)
                    .findAll()
                    .sort("mName", Sort.ASCENDING)

                contactsAdapter = ContactsAdapter(context, notSyncedContacts, true)
                contactsAdapter.onItemClickListener = { contact ->
                    val facebookPicketIntent =
                        Intent(context, FacebookPickerActivity::class.java).apply {
                            putExtra(FacebookPickerActivity.EXTRA_ID, contact.getId())
                        }
                    parentActivity.startActivityForResult(
                        facebookPicketIntent,
                        TabActivity.REQUEST_FACEBOOK_PICKER
                    )
                }
            }
            1 -> {
                val autoSyncedContacts = realmUi.where(Contact::class.java)
                    .isNotNull("related")
                    .equalTo("isManual", false)
                    .findAll()
                    .sort("mName", Sort.ASCENDING)

                contactsAdapter = ContactsAdapter(context, autoSyncedContacts, true)
                contactsAdapter.onItemClickListener = { contact ->
                    AlertDialog.Builder(context)
                        .setTitle(R.string.alert_cancel_auto_sync_title)
                        .setMessage(R.string.alert_cancel_auto_sync_message)
                        .setPositiveButton(android.R.string.ok) { dialogInterface, _ ->
                            realmUi.executeTransaction {
                                contact.related = null
                                contact.isManual = true
                                contact.synced = true
                            }
                            dialogInterface.dismiss()
                        }
                        .setNegativeButton(android.R.string.cancel) { dialogInterface, _ -> dialogInterface.dismiss() }
                        .create()
                        .show()
                }
            }
            2 -> {
                val manualContacts = realmUi.where(Contact::class.java)
                    .equalTo("isManual", true)
                    .findAll()
                    .sort("mName", Sort.ASCENDING)

                contactsAdapter = ContactsAdapter(context, manualContacts, true)
                contactsAdapter.onItemClickListener = { contact ->
                    AlertDialog.Builder(context)
                        .setTitle(R.string.alert_release_bond_title)
                        .setMessage(R.string.alert_release_bond_message)
                        //TODO: make another dialog/preference remembering if app should remove photo automatically
                        .setPositiveButton(android.R.string.ok) { dialogInterface, _ ->
                            realmUi.executeTransaction {
                                contact.related = null
                                contact.isManual = false
                                contact.synced = false
                            }
                            dialogInterface.dismiss()
                        }
                        .setNegativeButton(android.R.string.cancel) { dialogInterface, _ -> dialogInterface.dismiss() }
                        .create()
                        .show()
                }
            }
            else -> throw IllegalStateException("Shouldn't contain more than 3 tabs")
        }
        recyclerView.adapter = contactsAdapter
        return view
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
    }

    override fun getPageTitle(position: Int): CharSequence? = when (position) {
        0 -> parentActivity.getString(R.string.activity_tab_tab_not_synced)
        1 -> parentActivity.getString(R.string.activity_tab_tab_auto)
        2 -> parentActivity.getString(R.string.activity_tab_tab_manual)
        else -> super.getPageTitle(position)
    }

    companion object {

        private const val TAB_COUNT = 3
    }
}
